File: core/orchestrator.py
"""
The Director: runs one full simulated day across every channel adapter,
owns the agent swarm (spawn/deprecate per the 5-lead/day rule), routes
each qualified signal through core.pipeline, and feeds every outcome to
the Sentinel for drift tracking / hot-swap.

This is the orchestration layer described across the conversation this
project is based on, made concrete: Layer 1 (central command) -> Layer 2
(macro channel controllers) -> Layer 3 (platform workers) -> Layer 4
(dynamic sub-domain workers), with a cross-cutting sentinel plane.
"""
from __future__ import annotations

import random
import sqlite3
from collections import defaultdict
from dataclasses import dataclass, field

from core import db
from core.channels import all_adapters
from core.channels.base import ChannelAdapter
from core.filters import passes_layer0
from core.llm import LLMClient
from core.models import (
    AgentLayer,
    AgentNode,
    AgentStatus,
    ChannelType,
    now_iso,
    new_id,
)
from core.pipeline import run_pipeline
from core.sentinel import Sentinel
from core.strategist import StrategyLedger

SUBDOMAIN_SPAWN_THRESHOLD = 5   # qualifying leads/day -> spawn a dedicated L4 worker
SUBDOMAIN_DEPRECATE_THRESHOLD = 2  # below this -> deprecate the L4 worker

_FRICTION_PENALTY = {"LOW": 0.0, "MEDIUM": 0.15, "HIGH": 0.35}

# Per-agent fault bias: most spawned workers stay clean; a small minority
# are drawn as "degrading" (simulating a model/prompt-version drift on that
# one node) so the Reviewer/Sentinel/hot-swap loop has something real to
# catch instead of running forever with a spotless record.
DEGRADING_AGENT_PROBABILITY = 0.08
CLEAN_AGENT_FAULT_RATE = 0.02
DEGRADING_AGENT_FAULT_RATE = 0.45


@dataclass
class ChannelCycleSummary:
    channel_type: str
    platforms: int
    signals: int
    layer0_passed: int
    qualified: int
    bids_approved: int
    bids_rejected: int
    subdomain_agents_spawned: list[str] = field(default_factory=list)
    subdomain_agents_deprecated: list[str] = field(default_factory=list)
    channel_health_score: float = 0.0


@dataclass
class CycleSummary:
    day: int
    channels: list[ChannelCycleSummary]
    sentinel_events: int
    total_signals: int
    total_qualified: int
    total_bids_approved: int


class Orchestrator:
    def __init__(self, llm: LLMClient | None = None):
        self.llm = llm
        self.ledger = StrategyLedger()
        self.sentinel = Sentinel()
        # agent registry keyed by a stable identity string, so re-running
        # a cycle reuses the same agents instead of spawning duplicates
        self.agents: dict[str, AgentNode] = {}
        self.agent_fault_rate: dict[str, float] = {}

    # -- agent registry helpers -------------------------------------------------

    def _assign_fault_bias(self, agent_id: str) -> None:
        r = random.Random(agent_id)
        degrading = r.random() < DEGRADING_AGENT_PROBABILITY
        self.agent_fault_rate[agent_id] = DEGRADING_AGENT_FAULT_RATE if degrading else CLEAN_AGENT_FAULT_RATE

    def _get_or_spawn(self, key: str, layer: AgentLayer, role: str,
                       channel_type: ChannelType | None, sub_domain: str | None) -> AgentNode:
        if key in self.agents and self.agents[key].status != AgentStatus.DEPRECATED:
            return self.agents[key]
        agent = AgentNode(
            agent_id=new_id("agent"),
            layer=layer,
            role=role,
            channel_type=channel_type,
            sub_domain=sub_domain,
            status=AgentStatus.ACTIVE,
            model_signature="primary-extractor-v1",
        )
        self.agents[key] = agent
        self._assign_fault_bias(agent.agent_id)
        return agent

    # -- main cycle ---------------------------------------------------------

    def run_cycle(self, day: int, conn: sqlite3.Connection | None = None) -> CycleSummary:
        owns_conn = conn is None
        conn = conn or db.connect()
        db.init_schema(conn)

        central = self._get_or_spawn("L1:director", AgentLayer.L1_CENTRAL_COMMAND, "Central Director", None, None)
        db.save_agent(conn, central)

        channel_summaries: list[ChannelCycleSummary] = []
        events_before = len(self.sentinel.events)

        for adapter in all_adapters():
            summary = self._run_channel(conn, adapter, day)
            channel_summaries.append(summary)

        for rec in self.ledger.all_records():
            db.save_strategy(conn, rec)

        conn.commit()
        if owns_conn:
            conn.close()

        return CycleSummary(
            day=day,
            channels=channel_summaries,
            sentinel_events=len(self.sentinel.events) - events_before,
            total_signals=sum(c.signals for c in channel_summaries),
            total_qualified=sum(c.qualified for c in channel_summaries),
            total_bids_approved=sum(c.bids_approved for c in channel_summaries),
        )

    def _run_channel(self, conn: sqlite3.Connection, adapter: ChannelAdapter, day: int) -> ChannelCycleSummary:
        ct = adapter.channel_type
        l2_key = f"L2:{ct.value}"
        l2_agent = self._get_or_spawn(l2_key, AgentLayer.L2_MACRO_CHANNEL, f"{ct.value} Channel Controller", ct, None)
        db.save_agent(conn, l2_agent)

        signals = adapter.generate_daily_signals(day)
        subdomain_qualified_counts: dict[str, int] = defaultdict(int)
        subdomain_platform: dict[str, str] = {}

        layer0_passed = qualified = bids_approved = bids_rejected = 0

        for signal in signals:
            platform_key = f"L3:{ct.value}:{signal.platform}"
            worker = self._get_or_spawn(platform_key, AgentLayer.L3_PLATFORM_WORKER, f"{signal.platform} Worker", ct, None)

            responsible_key = platform_key
            if signal.sub_domain:
                sub_key = f"L4:{ct.value}:{signal.platform}:{signal.sub_domain}"
                if sub_key in self.agents and self.agents[sub_key].status == AgentStatus.ACTIVE:
                    responsible_key = sub_key
                    worker = self.agents[sub_key]

            fault_rate = self.agent_fault_rate.get(worker.agent_id, CLEAN_AGENT_FAULT_RATE)
            outcome = run_pipeline(signal, self.ledger, self.llm, fault_rate=fault_rate)

            l0 = passes_layer0(signal)
            db.save_signal(conn, signal, day, l0.passed, l0.reason)

            if outcome.lead is None:
                # nothing to attribute to the worker -- Layer 0 / qualification
                # rejections happen before any agent does meaningful work
                continue

            layer0_passed += 1
            qualified += 1
            if signal.sub_domain:
                subdomain_qualified_counts[f"{signal.platform}::{signal.sub_domain}"] += 1
                subdomain_platform[f"{signal.platform}::{signal.sub_domain}"] = signal.platform

            db.save_lead(conn, outcome.lead)
            if outcome.bid:
                db.save_bid(conn, outcome.bid)
                approved = outcome.bid.reviewer_status == "APPROVED"
                bids_approved += int(approved)
                bids_rejected += int(not approved)
                if approved:
                    self.ledger.record_outcome(outcome.bid.strategy_id, won=True)

                worker.leads_handled += 1
                # anomaly = the Writer needed at least one rewrite (Reviewer
                # caught something) OR never recovered within the rewrite
                # budget -- both are signal that this agent's output needed
                # correcting, which is exactly what the sentinel watches for.
                anomaly = (not approved) or (outcome.bid.rewrite_count > 0)
                if anomaly:
                    reason = f"reviewer_flagged_draft (rewrites={outcome.bid.rewrite_count}, final={outcome.bid.reviewer_status}): " + ",".join(outcome.bid.reviewer_notes)
                else:
                    reason = "ok"
                new_worker, event = self.sentinel.evaluate(worker, anomaly, reason)
                self.agents[responsible_key] = new_worker
                db.save_agent(conn, new_worker)
                if event:
                    db.save_sentinel_event(conn, event)
                    if event.event_type == "HOTSWAP_TRIGGERED":
                        # the quarantined predecessor is also persisted so the
                        # admin panel / UI can show what got pulled and why
                        db.save_agent(conn, worker)
                        self._assign_fault_bias(new_worker.agent_id)
            else:
                db.save_agent(conn, worker)

        spawned, deprecated = self._reconcile_subdomains(conn, ct, subdomain_qualified_counts, subdomain_platform)

        db.save_run_log(
            conn, day, ct.value, len(signals), layer0_passed, qualified,
            bids_approved, bids_rejected, now_iso(),
        )

        friction = adapter.friction_level()
        velocity_score = min(1.0, qualified / 20.0)
        health = round(max(0.0, velocity_score - _FRICTION_PENALTY.get(friction, 0.15)), 3)

        return ChannelCycleSummary(
            channel_type=ct.value,
            platforms=len(adapter.list_platforms()),
            signals=len(signals),
            layer0_passed=layer0_passed,
            qualified=qualified,
            bids_approved=bids_approved,
            bids_rejected=bids_rejected,
            subdomain_agents_spawned=spawned,
            subdomain_agents_deprecated=deprecated,
            channel_health_score=health,
        )

    def _reconcile_subdomains(self, conn: sqlite3.Connection, ct: ChannelType,
                               counts: dict[str, int], platform_of: dict[str, str]) -> tuple[list[str], list[str]]:
        spawned, deprecated = [], []
        for combo, count in counts.items():
            platform, sub_domain = combo.split("::", 1)
            key = f"L4:{ct.value}:{platform}:{sub_domain}"
            existing = self.agents.get(key)

            if count >= SUBDOMAIN_SPAWN_THRESHOLD and (existing is None or existing.status != AgentStatus.ACTIVE):
                agent = AgentNode(
                    agent_id=new_id("agent"), layer=AgentLayer.L4_SUBDOMAIN_WORKER,
                    role=f"{sub_domain} Micro-Worker", channel_type=ct, sub_domain=sub_domain,
                    status=AgentStatus.ACTIVE, model_signature="primary-extractor-v1",
                )
                self.agents[key] = agent
                self._assign_fault_bias(agent.agent_id)
                db.save_agent(conn, agent)
                spawned.append(f"{platform}/{sub_domain}")
            elif existing is not None and existing.status == AgentStatus.ACTIVE and count < SUBDOMAIN_DEPRECATE_THRESHOLD:
                existing.status = AgentStatus.DEPRECATED
                db.save_agent(conn, existing)
                deprecated.append(f"{platform}/{sub_domain}")
        return spawned, deprecated

File: iechm-leadgen-engine-c/src/orchestrator.c
#include "orchestrator.h"
#include "channels.h"
#include "filters.h"
#include "pipeline.h"
#include "rng.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define SUBDOMAIN_SPAWN_THRESHOLD 5
#define SUBDOMAIN_DEPRECATE_THRESHOLD 2
#define DEGRADING_AGENT_PROBABILITY 0.08
#define CLEAN_AGENT_FAULT_RATE 0.02
#define DEGRADING_AGENT_FAULT_RATE 0.45

static double friction_penalty(const char *level) {
    if (strcmp(level, "LOW") == 0) return 0.0;
    if (strcmp(level, "MEDIUM") == 0) return 0.15;
    if (strcmp(level, "HIGH") == 0) return 0.35;
    return 0.15;
}

/* -------------------------------------------------------- agent registry */

static AgentRegistryEntry *registry_find(Orchestrator *o, const char *key) {
    for (int i = 0; i < o->agent_registry.count; i++) {
        AgentRegistryEntry *e = vec_at(&o->agent_registry, i);
        if (strcmp(e->key, key) == 0) return e;
    }
    return NULL;
}

static void assign_fault_bias(Orchestrator *o, const char *agent_id) {
    Rng r;
    rng_init_str(&r, agent_id);
    bool degrading = rng_next_double(&r) < DEGRADING_AGENT_PROBABILITY;
    FaultRateEntry *e = vec_push(&o->fault_rates);
    xcpy(e->agent_id, ID_LEN, agent_id);
    e->fault_rate = degrading ? DEGRADING_AGENT_FAULT_RATE : CLEAN_AGENT_FAULT_RATE;
}

static double get_fault_rate(Orchestrator *o, const char *agent_id) {
    for (int i = 0; i < o->fault_rates.count; i++) {
        FaultRateEntry *e = vec_at(&o->fault_rates, i);
        if (strcmp(e->agent_id, agent_id) == 0) return e->fault_rate;
    }
    return CLEAN_AGENT_FAULT_RATE;
}

static void spawn_agent_into(AgentNode *agent, AgentLayer layer, const char *role,
                              bool has_ct, ChannelType ct, bool has_sub, const char *sub_domain) {
    memset(agent, 0, sizeof *agent);
    new_id(agent->agent_id, sizeof agent->agent_id, "agent");
    agent->layer = layer;
    xcpy(agent->role, sizeof agent->role, role);
    agent->has_channel_type = has_ct;
    agent->channel_type = ct;
    agent->has_sub_domain = has_sub;
    xcpy(agent->sub_domain, sizeof agent->sub_domain, has_sub ? sub_domain : "");
    agent->status = AS_ACTIVE;
    xcpy(agent->model_signature, sizeof agent->model_signature, "primary-extractor-v1");
    agent->drift_score = 0.0;
    agent->consecutive_flags = 0;
    agent->leads_handled = 0;
    now_iso(agent->spawned_at, sizeof agent->spawned_at);
}

static AgentNode *registry_get_or_spawn(Orchestrator *o, const char *key, AgentLayer layer, const char *role,
                                         bool has_ct, ChannelType ct, bool has_sub, const char *sub_domain) {
    AgentRegistryEntry *e = registry_find(o, key);
    if (e && e->agent.status != AS_DEPRECATED) return &e->agent;

    if (!e) {
        e = vec_push(&o->agent_registry);
        xcpy(e->key, AGENT_KEY_LEN, key);
    }
    spawn_agent_into(&e->agent, layer, role, has_ct, ct, has_sub, sub_domain);
    assign_fault_bias(o, e->agent.agent_id);
    return &e->agent;
}

void orchestrator_init(Orchestrator *o) {
    strategy_ledger_init(&o->ledger);
    sentinel_init(&o->sentinel);
    vec_init(&o->agent_registry, sizeof(AgentRegistryEntry));
    vec_init(&o->fault_rates, sizeof(FaultRateEntry));
}

/* --------------------------------------------------------- per-channel run */

typedef struct { char combo[SHORT_LEN * 2]; char platform[SHORT_LEN]; char sub_domain[SHORT_LEN]; int count; } ComboCount;
#define COMBOS_MAX 64

static ComboCount *combo_find_or_add(ComboCount *combos, int *n, const char *platform, const char *sub_domain) {
    char combo[SHORT_LEN * 2];
    snprintf(combo, sizeof combo, "%s::%s", platform, sub_domain);
    for (int i = 0; i < *n; i++) if (strcmp(combos[i].combo, combo) == 0) return &combos[i];
    if (*n >= COMBOS_MAX) return &combos[COMBOS_MAX - 1];
    ComboCount *c = &combos[(*n)++];
    xcpy(c->combo, sizeof c->combo, combo);
    xcpy(c->platform, sizeof c->platform, platform);
    xcpy(c->sub_domain, sizeof c->sub_domain, sub_domain);
    c->count = 0;
    return c;
}

static void join_notes_commas(char notes[][NOTE_LEN], int count, char *out, size_t outcap) {
    out[0] = '\0';
    size_t used = 0;
    for (int i = 0; i < count; i++) {
        if (i > 0) { xcpy(out + used, outcap - used, ","); used = strlen(out); }
        xcpy(out + used, outcap - used, notes[i]);
        used = strlen(out);
    }
}

static ChannelCycleSummary run_channel(Orchestrator *o, Store *st, const ChannelAdapter *adapter, int day) {
    ChannelType ct = adapter->channel_type;

    char l2_key[AGENT_KEY_LEN], l2_role[ROLE_LEN];
    snprintf(l2_key, sizeof l2_key, "L2:%s", channel_type_str(ct));
    snprintf(l2_role, sizeof l2_role, "%s Channel Controller", channel_type_str(ct));
    AgentNode *l2 = registry_get_or_spawn(o, l2_key, AL_L2_MACRO_CHANNEL, l2_role, true, ct, false, "");
    store_save_agent(st, l2);

    Vec signals;
    vec_init(&signals, sizeof(RawSignal));
    adapter->generate_daily_signals(day, &signals);

    ComboCount combos[COMBOS_MAX];
    int combos_n = 0;

    int layer0_passed = 0, qualified = 0, bids_approved = 0, bids_rejected = 0;

    for (int i = 0; i < signals.count; i++) {
        RawSignal *signal = vec_at(&signals, i);

        char platform_key[AGENT_KEY_LEN], platform_role[ROLE_LEN];
        snprintf(platform_key, sizeof platform_key, "L3:%s:%s", channel_type_str(ct), signal->platform);
        snprintf(platform_role, sizeof platform_role, "%s Worker", signal->platform);
        AgentNode *worker = registry_get_or_spawn(o, platform_key, AL_L3_PLATFORM_WORKER, platform_role, true, ct, false, "");

        char responsible_key[AGENT_KEY_LEN];
        xcpy(responsible_key, sizeof responsible_key, platform_key);
        if (signal->sub_domain[0]) {
            char sub_key[AGENT_KEY_LEN];
            snprintf(sub_key, sizeof sub_key, "L4:%s:%s:%s", channel_type_str(ct), signal->platform, signal->sub_domain);
            AgentRegistryEntry *existing = registry_find(o, sub_key);
            if (existing && existing->agent.status == AS_ACTIVE) {
                xcpy(responsible_key, sizeof responsible_key, sub_key);
                worker = &existing->agent;
            }
        }

        double fault_rate = get_fault_rate(o, worker->agent_id);
        PipelineOutcome outcome;
        run_pipeline(signal, &o->ledger, fault_rate, &outcome);

        Layer0Result l0 = passes_layer0(signal);
        store_save_signal(st, signal, day, l0.passed, l0.reason);

        if (!outcome.has_lead) continue;

        layer0_passed++;
        qualified++;
        if (signal->sub_domain[0]) {
            ComboCount *c = combo_find_or_add(combos, &combos_n, signal->platform, signal->sub_domain);
            c->count++;
        }

        store_save_lead(st, &outcome.lead);

        if (outcome.has_bid) {
            store_save_bid(st, &outcome.bid);
            bool approved = strcmp(outcome.bid.reviewer_status, "APPROVED") == 0;
            bids_approved += approved ? 1 : 0;
            bids_rejected += approved ? 0 : 1;
            if (approved) strategy_record_outcome(&o->ledger, outcome.bid.strategy_id, true);

            worker->leads_handled++;
            bool anomaly = (!approved) || (outcome.bid.rewrite_count > 0);
            char reason[TEXT_LEN];
            if (anomaly) {
                char notes_joined[NOTES_MAX * NOTE_LEN];
                join_notes_commas(outcome.bid.reviewer_notes, outcome.bid.reviewer_notes_count, notes_joined, sizeof notes_joined);
                snprintf(reason, sizeof reason, "reviewer_flagged_draft (rewrites=%d, final=%s): %s",
                         outcome.bid.rewrite_count, outcome.bid.reviewer_status, notes_joined);
            } else {
                xcpy(reason, sizeof reason, "ok");
            }

            SentinelEvalResult sr = sentinel_evaluate(&o->sentinel, worker, anomaly, reason);

            if (sr.replaced) {
                store_save_agent(st, &sr.replacement);
                store_save_agent(st, worker); /* quarantined predecessor, distinct agent_id row */
                assign_fault_bias(o, sr.replacement.agent_id);
                AgentRegistryEntry *slot = registry_find(o, responsible_key);
                if (slot) slot->agent = sr.replacement;
            } else {
                store_save_agent(st, worker);
            }
            if (sr.has_event) store_save_sentinel_event(st, &sr.event);
        } else {
            store_save_agent(st, worker);
        }
    }

    ChannelCycleSummary summary;
    memset(&summary, 0, sizeof summary);
    xcpy(summary.channel_type, sizeof summary.channel_type, channel_type_str(ct));
    int platform_count;
    adapter->list_platforms(&platform_count);
    summary.platforms = platform_count;
    summary.signals = signals.count;
    summary.layer0_passed = layer0_passed;
    summary.qualified = qualified;
    summary.bids_approved = bids_approved;
    summary.bids_rejected = bids_rejected;

    for (int i = 0; i < combos_n; i++) {
        ComboCount *c = &combos[i];
        char key[AGENT_KEY_LEN];
        snprintf(key, sizeof key, "L4:%s:%s:%s", channel_type_str(ct), c->platform, c->sub_domain);
        AgentRegistryEntry *existing = registry_find(o, key);

        if (c->count >= SUBDOMAIN_SPAWN_THRESHOLD && (!existing || existing->agent.status != AS_ACTIVE)) {
            char role[ROLE_LEN];
            snprintf(role, sizeof role, "%s Micro-Worker", c->sub_domain);
            AgentNode *spawned = registry_get_or_spawn(o, key, AL_L4_SUBDOMAIN_WORKER, role, true, ct, true, c->sub_domain);
            store_save_agent(st, spawned);
            if (summary.spawned_count < CHANNEL_SPAWN_LIST_MAX)
                snprintf(summary.spawned[summary.spawned_count++], ROLE_LEN, "%s/%s", c->platform, c->sub_domain);
        } else if (existing && existing->agent.status == AS_ACTIVE && c->count < SUBDOMAIN_DEPRECATE_THRESHOLD) {
            existing->agent.status = AS_DEPRECATED;
            store_save_agent(st, &existing->agent);
            if (summary.deprecated_count < CHANNEL_SPAWN_LIST_MAX)
                snprintf(summary.deprecated[summary.deprecated_count++], ROLE_LEN, "%s/%s", c->platform, c->sub_domain);
        }
    }

    char created_at[SHORT_LEN];
    now_iso(created_at, sizeof created_at);
    store_save_run_log(st, day, channel_type_str(ct), signals.count, layer0_passed, qualified,
                        bids_approved, bids_rejected, created_at);

    double velocity = qualified / 20.0;
    if (velocity > 1.0) velocity = 1.0;
    double health = velocity - friction_penalty(adapter->friction_level());
    if (health < 0.0) health = 0.0;
    summary.channel_health_score = health;

    free(signals.items);
    return summary;
}

void orchestrator_run_cycle(Orchestrator *o, Store *st, int day, CycleSummary *out) {
    memset(out, 0, sizeof *out);

    AgentNode *central = registry_get_or_spawn(o, "L1:director", AL_L1_CENTRAL_COMMAND, "Central Director", false, 0, false, "");
    store_save_agent(st, central);

    int events_before = o->sentinel.event_count;

    for (int i = 0; i < channels_count(); i++) {
        const ChannelAdapter *adapter = channels_get(i);
        out->channels[out->channels_count++] = run_channel(o, st, adapter, day);
    }

    for (int i = 0; i < o->ledger.count; i++) store_save_strategy(st, &o->ledger.records[i]);

    out->day = day;
    out->sentinel_events_this_cycle = o->sentinel.event_count - events_before;
    for (int i = 0; i < out->channels_count; i++) {
        out->total_signals += out->channels[i].signals;
        out->total_qualified += out->channels[i].qualified;
        out->total_bids_approved += out->channels[i].bids_approved;
    }
}
